app/models/file.js

```diff
@@ -1,32 +1,82 @@
-//native modules
-var path = require('path');
+// native modules
+const path = require('path');
 
 // 3rd party modules
-var mongoose = require('mongoose');
-var Schema = mongoose.Schema;
-
-var nconf = require('../../config');
-var file_provider = nconf.get('filedb');
-
-var FileSchema = new Schema({
-    //buffer limit 16MB when attached to document!
-    name: {type: String},
-    mime_type: {type: String},
-    base64: {type: String},
-    data: {type: Buffer},
-    size: {type: Number},
-    sha1: {type: String},
-    sha256: {type: String}
+const logger = require('winston');
+const mongoose = require('mongoose');
+
+// local module
+const nconf = require('../../config');
+const checksum = require('../tools/checksum.js');
+const fileProvider = nconf.get('filedb');
+
+const FileSchema = new mongoose.Schema({
+  // buffer limit 16MB when attached to document!
+  name: { type: String },
+  mime_type: { type: String },
+  encoding: { type: String, enum: ['raw', 'base64'], default: 'raw' },
+  data: { type: Buffer },
+  size: { type: Number },
+  sha1: { type: String },
+  sha256: { type: String }
+});
+FileSchema.set('toObject', { virtuals: true });
+
+FileSchema.virtual('hrefs').get(function () {
+  return (fileProvider && fileProvider !== 'mongodb' && this.sha1) ? path.join(fileProvider, this.sha1) : undefined;
 });
-FileSchema.set('toObject', {virtuals: true});
 
-FileSchema.virtual('hrefs').get(
-    function () {
-        var href;
-        if (file_provider && file_provider !== 'mongodb' && this.sha1) {
-            href = path.join(file_provider, this.sha1);
-        }
-        return href;
-    });
+FileSchema.methods.prepareDataForStorage = function () {
+  logger.info(`Preparing file (name: ${this.name}) for storage.`);
+
+  if (this.encoding === 'base64') {
+    logger.debug('Base64 file detected, storing data to a buffer.');
+    this.data = new Buffer(this.data, 'base64');
+  }
+
+  if (this.data) {
+    this.size = this.data.length;
+    // file.type = mimetype(file.name(
+    this.sha1 = checksum(this.data, 'sha1');
+    this.sha256 = checksum(this.data, 'sha256');
+  }
+};
+
+FileSchema.methods.storeInFileDB = function () {
+  // filedb is reuired here because it causes a circular dependency otherwise
+  const filedb = require('../tools/filedb.js'); // eslint-disable-line
```

Comment:
filedb and file model are a bit too tightly coupled
filedb needs File model so it can check that it has received a valid file before requests.
File model on the other hand needs filedb to perform file storing requests.

solutions I can think of
* filedb does not check whether it has received a valid file or not
* File model informs filedb what File to use after it has loaded and introduced itself to mongoose (might cause async issues)
* filedb asks mongoose for the model every time it needs to check whether it has received a proper file
* the one above used in the code /\

eslint is disabled because it really does not tolerate requires in the middle of the file.